### ISSUE TITLE
Question: How to evaluate models

### ISSUE THREAD

**Issue by aclifton314** [UKPLab/sentence-transformers#48]
Suppose I have a data set representative of my particular domain that I use to fine-tune a pretrained model on. What would be a sensible procedure to go about evaluating the two models? Said differently, how would I know if fine-tuning was worth it?

**Comment by nreimers** [comment]
Hi @aclifton314,
evaluation always depends on your task and is an important step. This repository aims for unsupervised tasks like clustering or semantic search. There, the use cases are too broad and it is hard to name one winning solution. Often, an approach (one sentence embeddings method) that works well for e.g. clustering does not work well for semantic search and vice versa. 

Hence, evaluation on your task with your specific data is a must.

For this, you must create some labeled data. For semantic search, you e.g. label some search queries with relevant and non-relevant documents. For clustering, you provide some gold cluster labels.

With these labels, you can then compute performances.

Best regards
Nils Reimers

**Comment by aclifton314** [comment]
@nreimers 
Thanks for your response.

Sticking with the semantic search task, if I backup a bit to fine-tuning a pretrained model, would I need a similar set of labeled data to fine-tune? And if using cosine similarity as the loss what would the labels be for the data used to further fine-tune, say, bert-base-nli-mean-tokens?

**Comment by nreimers** [comment]
Hi @aclifton314 ,
for semantic search, triplet loss can be quite interesting: There you must choose a triplet: anchor, positive example, negative example.

For semantic search,  this could be query, relevant sentence, non-relevant sentence. Often, it is possible to generate these triplet dataset automatically (i.e. from other data scoures you already have).

For example you do semantic search for stackoverflow. There, you have often links posted by users between threads. You can assume that a link indicates that two threads are similar. You construct the triplet dataset in this way: anchor thread, linked thread, random thread (without a link to the anchor thread).

I hope this helps. Creating datasets for tuning sentence embeddings is sadly the hardest part.

Best regards
Nils Reimers

**Comment by aclifton314** [comment]
@nreimers Thank you for your response. Makes perfect sense to me! Closing this issue out.
